grid2op/multi_agent/multiAgentEnv.py

```diff
@@ -0,0 +1,557 @@
+        self._action_domains = {k: {"sub_id": copy.deepcopy(v)} for k,v in action_domains.items()}
+        self.num_agents = len(action_domains)
+        
+        if observation_domains is not None:
+            # user specified an observation domain
+            self._is_global_obs : bool = False
+            if action_domains.keys() != observation_domains.keys():
+                raise("action_domains and observation_domains must have the same agents' names !")
+            self._observation_domains = {k: {"sub_id": copy.deepcopy(v)} for k,v in observation_domains.items()}
+            self._verify_domains(observation_domains)
+        else:
+            # no observation domain specified, so I assume it's a domain
+            # with full observability
+            self._is_global_obs : bool = True
+            self._observation_domains = None
+        
+        self.agents = list(action_domains.keys())
+        self.num_agents = len(self.agents)
+        
+        self.illegal_action_pen = illegal_action_pen
+        self.ambiguous_action_pen = ambiguous_action_pen
+        self.rewards = dict(
+            zip(
+                self.agents, [0. for _ in range(self.num_agents)]
+            )
+        )
+        self._cumulative_rewards = dict(
+            zip(
+                self.agents, [0. for _ in range(self.num_agents)]
+            )
+        )
+        self.observations = dict(
+            zip(
+                self.agents, [None for _ in range(self.num_agents)]
+            )
+        )
+        self.agent_order = AgentSelector(self.agents, agent_order_fn)
+        self.action_spaces = dict()
+        self.observation_spaces = dict()
+        self._build_subgrids()
+        self._build_action_spaces()
+        self._build_observation_spaces()
+        
+        
+    def reset(self) -> MADict:
+        self._cent_env.reset()
+        self._update_observations(_update_state=False)
+        return self.observations
+    
+    def step(self, action : ActionProfile) -> Tuple[MADict, MADict, MADict, MADict]:
+        raise NotImplementedError()
+    
+    def _build_subgrids(self):
+        
+        self._subgrids_cls = {
+            'action' : dict(),
+            'observation' : dict()
+        }
+        for agent_nm in self.agents : 
+            # action space
+            self._build_agent_domain(agent_nm, self._action_domains[agent_nm])
+            subgridcls = self._build_subgrid_cls_from_domain(self._action_domains[agent_nm])
+            self._subgrids_cls['action'][agent_nm] = subgridcls 
+            
+            # observation space
+            if self._observation_domains is not None:
+                self._build_agent_domain(agent_nm, self._observation_domains[agent_nm])
+                subgridcls = self._build_subgrid_cls_from_domain(self._observation_domains[agent_nm])
+                self._subgrids_cls['observation'][agent_nm] = subgridcls
+        
+    def _build_agent_domain(self, agent_nm, domain):
```

Comment:
it's more "build_subgrid_masks" now I believe

Reply:
Changed it